### PR TITLE
build: The Light Clangtastic

### PR DESCRIPTION
**Pull request by adamemerson**
Do you want to compile with Clang?

Perhaps you do! But for a while you couldn't, due to various infelicities and deviations from proper C++. This PR fixes those as well as some warnings.

### CODE REVIEW

**Comment by mattbenjamin** [comment]
looks good, going to try building

**Comment by mattbenjamin** [comment]
I'd say this is deeper than the changeset :)

[ 39%] Building CXX object src/CMakeFiles/common.dir/perfglue/disabled_stubs.cc.o
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.7.0 (tags/RELEASE_370/final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
clang: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/block_based_table_reader-fd50bf.cpp
clang: note: diagnostic msg: /tmp/block_based_table_reader-fd50bf.sh
clang: note: diagnostic msg:

********************
CMakeFiles/rocksdblib.dir/build.make:1526: recipe for target 'CMakeFiles/rocksdblib.dir/table/block_based_table_reader.cc.o' failed

File: CMakeLists.txt [hunk 1]
```diff
@@ -462,8 +462,8 @@ endif()
 
 # require minimally the bundled version
 find_package(Boost 1.61 COMPONENTS thread system regex random program_options date_time iostreams REQUIRED)
-include_directories(${Boost_INCLUDE_DIRS})
-include_directories(${PROJECT_BINARY_DIR}/include)
+include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
```
Comment (adamemerson) [inline CMakeLists.txt]: It's not required with Clang, but I'd want it with GCC anyway. Getting warnings from the Boost include files is just noise for us and distracts from keeping our own code clean.

(We ought not get used to paged of warnings scrolling by.)

File: src/key_value_store/kv_flat_btree_async.cc [hunk 2]
```diff
@@ -1120,7 +1120,7 @@ int KvFlatBtreeAsync::cleanup(const index_data &idata, const int &error) {
     //all changes were created except for updating the index and possibly
     //deleting the objects. roll forward.
     vector<pair<pair<int, string>, librados::ObjectWriteOperation*> > ops;
-    librados::ObjectWriteOperation owos[idata.to_delete.size() + 1];
+    vector<librados::ObjectWriteOperation> owos(idata.to_delete.size() + 1);
```
Comment (adamemerson) [inline src/key_value_store/kv_flat_btree_async.cc]: If we really wanted we could make a variable-length fixed-size array (i.e. its size is determined by runtime) that lives on the stack and can hold objects like this, but that's a task for another time.

**Comment by mattbenjamin** [comment]
Ok, looks like clang++ is just running out of memory faster than g++ does, so I can't build at -j 22 like I usually do :)

**Comment by mattbenjamin** [comment]
Hmm.  Actually not: having trouble w/rocksdb compile:

[ 24%] Building CXX object CMakeFiles/rocksdblib.dir/util/file_reader_writer.cc.o
clang: error: unable to execute command: Segmentation fault (core dumped)
clang: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.7.0 (tags/RELEASE_370/final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
clang: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/block_based_table_reader-fe6739.cpp
clang: note: diagnostic msg: /tmp/block_based_table_reader-fe6739.sh
clang: note: diagnostic msg: 

********************
CMakeFiles/rocksdblib.dir/build.make:1526: recipe for target 'CMakeFiles/rocksdblib.dir/table/block_based_table_reader.cc.o' failed
make[6]: *** [CMakeFiles/rocksdblib.dir/table/block_based_table_reader.cc.o] Error 254
make[6]: *** Waiting for unfinished jobs....
[ 11%] Built target cls_numops_client
[ 12%] Built target cls_rbd
[ 12%] Built target cls_rbd_client
CMakeFiles/Makefile2:4321: recipe for target 'CMakeFiles/rocksdblib.dir/all' failed
make[5]: *** [CMakeFiles/rocksdblib.dir/all] Error 2
CMakeFiles/Makefile2:4333: recipe for target 'CMakeFiles/rocksdblib.dir/rule' failed
make[4]: *** [CMakeFiles/rocksdblib.dir/rule] Error 2
Makefile:1481: recipe for target 'rocksdblib' failed

**Comment by adamemerson** [comment]
It works on Clang 3.8 at least. I wonder if we could get the FreeBSD Person to try it.

**Comment by wjwithagen** [comment]
@adamemerson 
I'm using 3.8.0 on FreeBSD ( and sometimes the 4.0-beta, but that does not help in most cases) 
3.7.0 has proven to give errors in some of the tests. 

Now the major number of changes you have are against RBD, and I've excluded building that atm. So those will not be tested right now.
Also the embedded Boost is currently not used as there are some incompatibilities that are in need of figuring out.
OTOH things like std::move and the initializers are really things that I see flying by on a regular basis.

Now the trick question is how I can easily get this PR into my wjwithagen:wip-wjw-freebsd-cmake tree?
And then we will find out if things work.

**Comment by adamemerson** [comment]
The simplest might be to check out a copy of this branch, rebase -i it onto your wip-wjw-freebsd-cmake branch and just drop every commit before f59fbc7983039db63123089401a99d4ee88b9341 so you don't pull in the stuff from master along with it.

**Comment by wjwithagen** [comment]
@adamemerson 
I know how to to rebase my own stuff with master, or rebase for fixing and squashing.
But this is a new one.
I can checkout your fork  into Ceph/adamemerson/ceph and checkout your wip.
But what is the rebase  incantation to get it into Ceph/work/ceph?

**Comment by adamemerson** [comment]
If you mean two different repositories, I don't think you can rebase it. You can use format-patch to make a patch series and just git am to apply them on the other repository.

File: src/rgw/rgw_rest_s3.h [hunk 3]
```diff
@@ -761,8 +761,8 @@ class RGWGetPolicyV2Extractor:public RGWS3V2AuthEngine::Extractor {
 
 public:
   RGWGetPolicyV2Extractor(std::string access_key_id, std::string signature) {
-    access_key_id = std::move(access_key_id),
-    signature = std::move(signature);
+    this->access_key_id = std::move(access_key_id),
```
Comment (tchaikov) [inline src/rgw/rgw_rest_s3.h]: could just put `std::move(...)` in the initialization list, like
```c++
RGWGetPolicyV2Extractor(std::string access_key_id, std::string signature)
 : access_key_id(std::move(access_key_id) ...
```

**Comment by wjwithagen** [comment]
@adamemerson 
LGTM for the parts that I compile
It compiles with a few errors less, especially the std::move stuff is rather verbose.

File: src/msg/async/Stack.cc [hunk 4]
```diff
@@ -31,13 +31,12 @@
...
 
-void NetworkStack::add_thread(unsigned i, std::function<void ()> &thread)
+std::function<void ()> NetworkStack::add_thread(unsigned i)
 {
   Worker *w = workers[i];
-  thread = std::move(
-    [this, w]() {
-      const uint64_t EventMaxWaitUs = 30000000;
-      w->center.set_owner();
+  std::function<void ()> thread = [this, w]() {
```
Comment (adamemerson) [inline src/msg/async/Stack.cc]: One could, it wouldn't hurt anything, this seemed clearer. If we were using C++14 we could just set the return type to 'auto' and return the lambda without packing it into a std::function for the caller to do what they like with.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2016-11-13_12:19:00-rados-wip-kefu-testing---basic-mira/

tests passed.

**Pull request merged**